### PR TITLE
useContainerQuery hook

### PR DESCRIPTION
**Pull request by mayank23**
Hi, this is my PR to add the useContainerQuery Hook. Added Readme update + tests which are all were passing locally. 

I kept getting the `ResizeObserverLoop` Limit exceeded error when running tests, so added the fix in which schedules the observer callback in an animation frame. I used the PR code from https://github.com/react-container-query/react-container-query/pull/77 . See https://github.com/react-container-query/react-container-query/issues/70 for additional info.

Wasn't able to get some of the typescripts types working correctly for useState, useCallback - on those lines I used `ts-ignore` for now

I ran the latest code in codesandbox and it was working fine. https://codesandbox.io/s/zen-violet-9kbm3

Please let me know if anything needs to be updated/changed

cc: @pixelbandito

### CODE REVIEW

File: test/server/index.js [hunk 1]
```diff
@@ -85,4 +85,4 @@ describe('applyContainerQuery', function () {
       .toBe('<div>{&quot;mobile&quot;:true,&quot;desktop&quot;:false}</div>');
   });
 
-});
+});
```
Comment (pixelbandito) [inline test/server/index.js:88]: I ran the code locally, and added a couple of tests. Good news: the code passed without modification.

Would you add the server rendering tests to your PR branch?

Here's what I did, pretty minimal, but it works if you want to copy-paste it.

```js
describe('useContainerQuery', function () {

  const query = {
    mobile: { maxWidth: 399 },
    desktop: { minWidth: 400 }
  };

  function renderContainerToStaticMarkup({ initialSize }) {
    const Container = () => {
      const [params, containerRef] = useContainerQuery(query, initialSize);

      return React.createElement(
        'div',
        {
          ref: containerRef
        },
        JSON.stringify(params)
      );
    };

    return renderToStaticMarkup(React.createElement(Container));
  }

  it('no matches with no initial widths', function () {
    const output = renderContainerToStaticMarkup({});
    expect(output).toBe('<div>{}</div>');
  });

  it('sets "mobile" to true at small sizes', function () {
    const zeroWidthOutput = renderContainerToStaticMarkup({ initialSize: { width: 399 } });
    const maximumMobileWidthOutput = renderContainerToStaticMarkup({ initialSize: { width: 0 } });

    expect(zeroWidthOutput).toBe('<div>{&quot;mobile&quot;:true,&quot;desktop&quot;:false}</div>');
    expect(maximumMobileWidthOutput).toBe('<div>{&quot;mobile&quot;:true,&quot;desktop&quot;:false}</div>');
  });

  it('sets "desktop" to true at large sizes', function () {
    const output = renderContainerToStaticMarkup({ initialSize: { width: 400 } });
    expect(output).toBe('<div>{&quot;mobile&quot;:false,&quot;desktop&quot;:true}</div>');
  });
});
```

File: src/index.tsx [hunk 2]
```diff
@@ -5,6 +5,49 @@ import {Props, State, Params, Query, Size} from './interfaces';
...
+  // setup a ref callback
+  // (end user) attaches that ref callback to a container element
+  // @ts-ignore
+  const [params, setParams] = React.useState(() => {
+    if (!initialSize) {
+      return {};
+    }
+    return matchQueries(query)(initialSize);
+  });
+  // @ts-ignore
```
Comment (pixelbandito) [inline src/index.tsx:26]: The ts-ignore is ugly, but I think we'll have to update typescript and/or types/react, types/react-dom to get it to work.

That should probably be outside the scope of this change, so I'll try to have a look at it later.

**Review by pixelbandito — CHANGES_REQUESTED** [review]
Changes requested:
1. Revert the update to .gitignore
2. Remove a comment referring to the PR
3. Add server tests

Otherwise, this looks very nice. 👏 👏 👏

**Comment by mayank23** [comment]
@pixelbandito , made changes based on the feedback 😄 

Also, speaking of the server rendering, since container query core is updated schedule the resize observer callbacks using `window.requestAnimationFrame` -  could this be a breaking change since `window` wouldn't be available in server side rendering I think? If so, maybe this release should be a major version bump 🤔

However I see that ResizeObserverLite also uses window.ResizeObserver or switches to using element-resize-observer which both would only work in the browser and not server side I think 
https://github.com/react-container-query/resize-observer-lite/blob/master/src/index.ts#L10-L16

TLDR: not sure if the existing react-container-query would actually work with server side rendering

**Comment by pixelbandito** [comment]
> TLDR: not sure if the existing react-container-query would actually work with server side rendering

I didn't write the original code, but afaik, the server-side rendering doesn't do any resize watching. It's only responsible for rendering using the initial width argument to the hook or component.

The tests seem to work as far as that goes, so I think it's ok.

**Comment by mayank23** [comment]
Alright great! Should be good to go then 😄 🥳
